logging/google/cloud/logging/_shutdown.py

```diff
@@ -0,0 +1,131 @@
+            for i in range(0, len(payload), MAX_PAYLOAD_SIZE)]
+
+
+def _write_stacktrace_log(client, traces):
+    """Writes the trace logs to the appropriate GAE resource in Stackdriver.
+
+    :type client: `google.cloud.logging.Client`
+    :param client: Stackdriver logging client.
+
+    :type traces: str
+    :param traces: String containing the stacktrace info to write to
+                   Stackdriver logging.
+    """
+    gae_version = _get_gae_version()
+    gae_service = _get_gae_service()
+    gae_instance = _get_gae_instance()
+
+    text_payload = '{}\nThread traces\n{}'.format(gae_instance, traces)
+
+    logger_name = 'projects/{}/logs/{}'.format(
+        client.project, 'appengine.googleapis.com%2Fapp.shutdown')
+
+    resource = {'type': 'gae_app', 'labels': {'project_id': client.project,
+                                              'version_id': gae_version,
+                                              'module_id': gae_service}}
+
+    labels = {
+        'appengine.googleapis.com/version_id': gae_version,
+        'compute.googleapis.com/resource_type': 'instance',
+        'appengine.googleapis.com/instance_name': gae_instance,
+        'appengine.googleapis.com/module_id': gae_service,
+    }
+
+    split_payloads = _split_entry(text_payload.encode('utf-8'))
+    entries = [{'text_payload': payload} for payload in split_payloads]
+
+    client.logging_api.write_entries(
+        entries, logger_name=logger_name, resource=resource, labels=labels)
+
+
+def _is_on_appengine():
+    """Returns True if the environment is detected as App Engine flexible."""
+    return (os.getenv(_APPENGINE_FLEXIBLE_ENV_VM) or os.getenv(
+        _APPENGINE_FLEXIBLE_ENV_FLEX))
+
+
+def _report_stacktraces(
+        client, signal_, frame):  # pylint: disable=unused-argument
+    """Reports the stacktraces of all active threads to Stackdriver Logging.
+
+    :type client: `google.cloud.logging.Client`
+    :param client: Stackdriver logging client.
+
+    :type signal_: int
+    :param signal_: Signal number.
+
+    :type frame: frame object
+    :param frame: The current stack frame.
+    """
+    traces = ''
+    for thread_id, stack in sys._current_frames().items():
+        traces += '\n# ThreadID: {}'.format(thread_id)
+        for filename, lineno, name, _ in traceback.extract_stack(stack):
+            traces += 'File: {}, line {}, in {}'.format(
+                filename, lineno, name)
+
+    _write_stacktrace_log(client, traces)
+
+
+def setup_stacktrace_crash_report(client):
+    """Installs a SIGTERM handler to log stack traces to Stackdriver.
+
+    :type client: `google.cloud.logging.Client`
+    :param client: Stackdriver logging client.
+    """
+    if not _is_on_appengine():
+        raise RuntimeError('Shutdown reporting is only supported on App '
+                           'Engine flexible environment.')
+    signal.signal(
+        signal.SIGTERM, functools.partial(_report_stacktraces, client))
```

Comment:
I'm on the fence on how to feel about a "very specific use case" as it pertains to this library. Would be interested in what @jonparrott and @lukesneeringer think.

:+1: on the `SIGTERM` explanation, though if it were my application I'd feel uneasy about the reliability of this (i.e. it may not always trigger for various reasons)